### PR TITLE
enforce single Language functions per function app

### PR DESCRIPTION
**Pull request by pragnagopa**
- If Functions_Worker_Runtime is set only load functions for that language
- If Functions_Worker_Runtime is not set, show error if more than one language fumctions are found

### CODE REVIEW

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 1]
```diff
@@ -688,6 +706,15 @@ internal Collection<FunctionDescriptor> GetFunctionDescriptors(IEnumerable<Funct
...
 
+        internal static bool FunctionsSupported(IEnumerable<FunctionMetadata> functions)
+        {
+            if (functions != null && functions.Any())
+            {
+                return functions.Where(funcMetadata => (funcMetadata.ScriptType != functions.ElementAt(0).ScriptType) || (funcMetadata.Language != functions.ElementAt(0).Language)).Any() == false;
+            }
+            return true;
+        }
+
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Why do we need to check both script type and language? Also wondering if it might be less confusing for this method to be implemented as
```csharp
return functions.Select(f => f.Language).Distinct().Count() <= 1
```

Comment (paulbatum) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Also, a better name might be "IsSingleLanguage"

**Comment by paulbatum** [comment]
Also want to double check - are these changes enough to make it so that things will work if a developer runs the webhost against the samples folder? Or are we going to require the dev to set the env variable to 'any'?

**Comment by pragnagopa** [comment]
Yes. To debug samples, we now need to set FUNCTIONS_WORKER_RUNTIME to any

File: sample/host.json [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
+      "routePrefix": "api",
+      "maxConcurrentRequests": 5,
+      "maxOutstandingRequests": 30
     },
-    "swagger": {
-        "enabled": true
+    "queues": {
+      "visibilityTimeout": "00:00:10"
     },
-    "extensions": {
```
Comment (fabiocav) [inline sample/host.json]: Why the changes? These had all been move under the `extensions`, which is where they should be.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 3]
```diff
@@ -465,31 +467,40 @@ private async Task InitializeWorkersAsync()
...
+            _functionDispatcher = new FunctionDispatcher(EventManager, server, channelFactory, _workerConfigs);
+
+            _eventSubscriptions.Add(EventManager.OfType<WorkerProcessErrorEvent>()
+                .Subscribe(evt =>
+                {
+                    HandleHostError(evt.Exception);
+                }));
+        }
+
+        internal static IEnumerable<WorkerConfig> GetWorkerConfigs(string language, ILogger logger)
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Can we move this out of `ScriptHost`? We're trying to reduce the mixed logic in this class.

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 4]
```diff
@@ -114,6 +116,18 @@ internal static FunctionMetadata ReadFunctionMetadata(string scriptDir, ICollect
                 }
                 else if (functionMetadata != null)
                 {
+                    if (functionMetadata.ScriptType == ScriptType.Unknown)
+                    {
+                        var workerConfig = workerConfigs.First(config => config.Extensions.Contains(Path.GetExtension(functionMetadata.ScriptFile)));
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: Wouldn't you have to use FirstOrDefault for the behavior you're expecting here? This will throw if you don't have a match.

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 5]
```diff
@@ -114,6 +116,18 @@ internal static FunctionMetadata ReadFunctionMetadata(string scriptDir, ICollect
                 }
                 else if (functionMetadata != null)
                 {
+                    if (functionMetadata.ScriptType == ScriptType.Unknown)
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: Can we move this logic to the `ParseScriptType` method so that it's all in the same place?

File: test/WebJobs.Script.Tests/ScriptHostTests.cs [hunk 6]
```diff
@@ -1031,6 +1031,56 @@ public void UpdateProxyName(string proxyName)
             Assert.Equal("myproxy", ProxyMetadataManager.NormalizeProxyName(proxyName));
         }
 
+        [Fact]
+        public void FunctionsSupported_Returns_True()
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/ScriptHostTests.cs]: nit: the method name here should be something like `IsSingleLanguage_WithSupportedLanguages_Returns_True`

File: test/WebJobs.Script.Tests/ScriptHostTests.cs [hunk 7]
```diff
@@ -1031,6 +1031,56 @@ public void UpdateProxyName(string proxyName)
...
+            };
+            IEnumerable<FunctionMetadata> functionsList = new Collection<FunctionMetadata>()
+            {
+                func1, func2
+            };
+            Assert.True(ScriptHost.IsSingleLanguage(functionsList));
+        }
+
+        [Fact]
+        public void FunctionsSupported_Returns_False()
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/ScriptHostTests.cs]: Same comment about the method name

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 8]
```diff
@@ -688,6 +706,15 @@ internal Collection<FunctionDescriptor> GetFunctionDescriptors(IEnumerable<Funct
             return functionDescriptors;
         }
 
+        internal static bool IsSingleLanguage(IEnumerable<FunctionMetadata> functions)
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: This can probably move elsewhere. We already have a few classes using this. `Utility` perhaps?

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 9]
```diff
@@ -465,31 +467,40 @@ private async Task InitializeWorkersAsync()
...
+            _eventSubscriptions.Add(EventManager.OfType<WorkerProcessErrorEvent>()
+                .Subscribe(evt =>
+                {
+                    HandleHostError(evt.Exception);
+                }));
+        }
+
+        internal static IEnumerable<WorkerConfig> GetWorkerConfigs(string language, ILogger logger)
+        {
+            var configFactory = new WorkerConfigFactory(ScriptSettingsManager.Instance.Configuration, logger);
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Let's please avoid new use of `ScriptSettingsManager`. If we must use configuration, we should just add that dependency to the constructor, but ideally, this config factory wouldn't be needed and `WorkerConfig` is just passed in as an `IOptions<WorkerConfig>`. or a collection of them.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 10]
```diff
@@ -465,31 +467,40 @@ private async Task InitializeWorkersAsync()
...
             var providers = new List<IWorkerProvider>();
-            if (!string.IsNullOrEmpty(_language))
+            // TODO: pgopa remove any flag
+            if (language == "any")
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: We should consider setting this automatically when running in development mode.

File: sample/host.json (hunk 2 above)
Comment (pragnagopa) [reply sample/host.json]: Will undo. Did not intend to make this change.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 9 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Sure will do.

**Comment by pragnagopa** [comment]
-  issue https://github.com/Azure/azure-functions-host/issues/3279  tracks moving worker config to DI.
- issue https://github.com/Azure/azure-functions-host/issues/3278 tracks tests refactoring

File: src/WebJobs.Script.WebHost/Startup.cs [hunk 11]
```diff
@@ -41,6 +41,7 @@ public void Configure(IApplicationBuilder app, IApplicationLifetime applicationL
             {
                 loggerFactory.AddConsole(LogLevel.Trace, true);
                 app.UseDeveloperExceptionPage();
+                Environment.SetEnvironmentVariable("FUNCTIONS_WORKER_RUNTIME", "any");
```
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/Startup.cs]: @fabiocav - Added environment variable here to help with F5 scenario.  Will address moving this out to IScriptWebHostEnvironment as part of https://github.com/Azure/azure-functions-host/issues/3278

File: src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs [hunk 12]
```diff
@@ -136,7 +134,8 @@ public async Task<(bool, FunctionMetadataResponse)> TryGetFunction(string name,
         {
             // TODO: DI (FACAVAL) Follow up with ahmels - Since loading of function metadata is no longer tied to the script host, we
             // should be able to inject an IFunctionMedatadaManager here and bypass this step.
-            var functionMetadata = FunctionMetadataManager.ReadFunctionMetadata(Path.Combine(_config.RootScriptPath, name), null, new Dictionary<string, ICollection<string>>(), fileSystem: FileUtility.Instance);
+            var workerConfigs = ScriptHost.GetWorkerConfigs("TODO", _logger);
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs]: This needs to be addressed, right?

File: src/WebJobs.Script.WebHost/Startup.cs (hunk 11 above)
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Startup.cs]: I think we can remove it for now. We just need to document it.

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 13]
```diff
@@ -279,5 +281,24 @@ private static ScriptType ParseScriptType(string scriptFilePath)
...
+                if (workerConfig != null)
+                {
+                    return workerConfig.Language;
+                }
+                return string.Empty;
+            }
+            return scriptType.ToString();
+        }
+
+        private static string GetScriptFileExtension(string scriptFilePath)
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: nit: Remove this method?

File: src/WebJobs.Script/Host/IFunctionMetadataManager.cs [hunk 14]
```diff
@@ -12,5 +13,7 @@ public interface IFunctionMetadataManager
         ImmutableDictionary<string, ImmutableArray<string>> Errors { get; }
 
         ImmutableArray<FunctionMetadata> Functions { get; }
+
+        IEnumerable<WorkerConfig> LanguageWorkerConfigs { get; set; }
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/IFunctionMetadataManager.cs]: Can we make this consistent with the other properties and expose an `ImmutableArray` here?

File: src/WebJobs.Script/Description/FunctionMetadata.cs [hunk 15]
```diff
@@ -36,6 +36,8 @@ public FunctionMetadata()
 
         public ScriptType ScriptType { get; set; }
 
+        public string Language { get; set; }
```
Comment (mathewc) [inline src/WebJobs.Script/Description/FunctionMetadata.cs]: The enum above ScriptType already captures the language - it will be confusing to have two

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 16]
```diff
@@ -152,7 +154,9 @@ internal static void ValidateName(string name, bool isProxy = false)
             }
 
             // determine the script type based on the primary script file extension
-            functionMetadata.ScriptType = ParseScriptType(functionMetadata.ScriptFile);
+            string extension = GetScriptFileExtension(functionMetadata.ScriptFile);
+            functionMetadata.ScriptType = ParseScriptType(extension);
```
Comment (mathewc) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: We need to consolidate the existing ScriptType and Language concepts

File: src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs [hunk 17]
```diff
@@ -136,7 +134,8 @@ public async Task<(bool, FunctionMetadataResponse)> TryGetFunction(string name,
...
+            var functionMetadata = FunctionMetadataManager.ReadFunctionMetadata(Path.Combine(_config.RootScriptPath, name), null, workerConfigs, new Dictionary<string, ICollection<string>>(), fileSystem: FileUtility.Instance);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs]: Rather than pass in the loose collection of worker configs, perhaps it would be cleaner to pass in an ILanguageWorkerService service that provides services like extension => language mappings, etc.? That would allow us to keep all such logic centralized.

File: src/WebJobs.Script/Description/FunctionMetadata.cs (hunk 15 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Description/FunctionMetadata.cs]: Tracks removing scriptType enum https://github.com/Azure/azure-functions-host/issues/3280

Comment (mathewc) [reply src/WebJobs.Script/Description/FunctionMetadata.cs]: I'd rather we just do this now rather than add a new duplicate property and incur debt and confusion. Shouldn't be that hard, right? Or perhaps just do that in a separate change and rebase this on that?

File: src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs (hunk 17 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs]: Don't really need DI - I'm suggesting that you encapsulate these new operations in a simple service implementation in this change set and pass that around, rather than making WorkerConfig public and passing those around.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 18]
```diff
@@ -653,6 +660,13 @@ internal Collection<FunctionDescriptor> GetFunctionDescriptors(IEnumerable<Funct
             Collection<FunctionDescriptor> functionDescriptors = new Collection<FunctionDescriptor>();
             var httpFunctions = new Dictionary<string, HttpTriggerAttribute>();
 
+            // TODO: pgopa remove any flag after refactoring tests to run in groups
```
[the remark this answers is not in the document]
Comment (mathewc) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Got it - I was sort of hoping that we'd retain the ability to run multi-language, for us as well as power users. I agree that if we're doing hard enforcement then we should update the tests refactoring them per language.

File: src/WebJobs.Script/Rpc/LanguageWorkerService.cs [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Microsoft.Azure.WebJobs.Script.Rpc
+{
+    internal class LanguageWorkerService : ILanguageWorkerService, IConfigureOptions<ScriptJobHostOptions>
+    {
+        private IEnumerable<WorkerConfig> _languageWorkerConfigs;
+
+        internal LanguageWorkerService(IConfiguration rootConfig)
+        {
+            ILoggerFactory loggerFactory = NullLoggerFactory.Instance;
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/LanguageWorkerService.cs]: @brettsam / @fabiocav  - Please let me know if there is a way to get startUp logger here

**Comment by pragnagopa** [comment]
- Replaced FunctionMetadata.ScriptType with FunctionMetadata.Language
- WorkerConfigs are now available as part of script options.

File: src/WebJobs.Script/Config/ScriptHostOptionsSetup.cs [hunk 20]
```diff
@@ -24,6 +28,7 @@ public ScriptHostOptionsSetup(IConfiguration configuration, IEnvironment environ
             _configuration = configuration;
             _environment = environment;
             _applicationHostOptions = applicationHostOptions;
+            _languageWorkerService = new LanguageWorkerService(configuration);
```
Comment (mathewc) [inline src/WebJobs.Script/Config/ScriptHostOptionsSetup.cs]: Rather than newing up the service here and assigning configs to ScriptOptions, you should add it as a service to ScriptHostBuilderExtensions, then it will be available everywhere:

```csharp
services.AddSingleton<ILanguageWorkerService, LanguageWorkerService>();
```

Then you don't have to stuff these configs on ScriptJobHostOptions where they don't belong.

File: src/WebJobs.Script/Rpc/LanguageWorkerService.cs (hunk 19 above)
Comment (mathewc) [reply src/WebJobs.Script/Rpc/LanguageWorkerService.cs]: I didn't see where you were adding this to the builder (see my comment above). If you did, then you'd get an ILoggerFactory from DI

File: src/WebJobs.Script/Rpc/LanguageWorkerService.cs [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+
+            var configFactory = new WorkerConfigFactory(rootConfig, loggerFactory.CreateLogger("Host.LanguageWorkerService"));
+            var providers = new List<IWorkerProvider>();
+            providers.AddRange(configFactory.GetWorkerProviders(NullLogger.Instance));
+            _languageWorkerConfigs = configFactory.GetConfigs(providers);
+        }
+
+        public IEnumerable<WorkerConfig> LanguageWorkerConfigs => _languageWorkerConfigs;
+
+        public void Configure(ScriptJobHostOptions options)
```
Comment (mathewc) [inline src/WebJobs.Script/Rpc/LanguageWorkerService.cs]: I don't think you need the IConfigureOptions<ScriptJobHostOptions> implementation. What is the goal?

File: src/WebJobs.Script/Rpc/LanguageWorkerService.cs [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            ILoggerFactory loggerFactory = NullLoggerFactory.Instance;
+
+            var configFactory = new WorkerConfigFactory(rootConfig, loggerFactory.CreateLogger("Host.LanguageWorkerService"));
+            var providers = new List<IWorkerProvider>();
+            providers.AddRange(configFactory.GetWorkerProviders(NullLogger.Instance));
+            _languageWorkerConfigs = configFactory.GetConfigs(providers);
+        }
+
+        public IEnumerable<WorkerConfig> LanguageWorkerConfigs => _languageWorkerConfigs;
```
Comment (mathewc) [inline src/WebJobs.Script/Rpc/LanguageWorkerService.cs]: I don't think you need to expose the raw configs, right? Don't you just need a new `bool TryGetLanguage(string extension, out string language)` that does the mapping of extension to language for you? That way the internal details stay internal, calling code is simplified, and WorkerConfig can stay internal.

File: src/WebJobs.Script/Description/DotNet/Compilation/DotNetCompilationServiceFactory.cs [hunk 23]
```diff
@@ -53,19 +53,19 @@ internal static void SetOptimizationLevel(OptimizationLevel? level)
...
-                case ScriptType.FSharp:
+                case "FSharp":
                     return new FSharpCompilationService(metadata, OptimizationLevel, _loggerFactory);
-                case ScriptType.DotNetAssembly:
+                case "DotNetAssembly":
                     return new RawAssemblyCompilationService();
                 default:
                     throw new NotSupportedException(string.Format(CultureInfo.InvariantCulture,
-                        "The script type {0} is not supported by the {1}", scriptType, typeof(DotNetCompilationServiceFactory).Name));
+                        "The script type {0} is not supported by the {1}", language, typeof(DotNetCompilationServiceFactory).Name));
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/Compilation/DotNetCompilationServiceFactory.cs]: Fix text "The language {0}..."

Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/Compilation/DotNetCompilationServiceFactory.cs]: I wonder if this text shouldn't be updated. I was thinking "The language {0} is not supported..." but "DotNetAssembly" isn't really a language. Hmmm.

File: src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilationService.cs [hunk 24]
```diff
@@ -1,36 +0,0 @@
...
-using System;
-using System.Collections.Generic;
-using System.IO;
-using System.Linq;
-using System.Threading.Tasks;
-using Microsoft.Azure.WebJobs.Script.Config;
-
-namespace Microsoft.Azure.WebJobs.Script.Description.Node.TypeScript
-{
-    public class TypeScriptCompilationService : ICompilationService<IJavaScriptCompilation>
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilationService.cs]: So we don't support type script anymore?

File: src/WebJobs.Script/Description/ScriptTypes.cs [hunk 25]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public class ScriptTypes
```
Comment (mathewc) [inline src/WebJobs.Script/Description/ScriptTypes.cs]: Please move this to Description/DotNet folder to make it clear that this is no longer intended to be a general list of supported languages. You might rename the class DotNetLanguage to make this clearer?

Comment (mathewc) [reply src/WebJobs.Script/Description/ScriptTypes.cs]: Also, I don't think you need the Unknown value?

File: src/WebJobs.Script/Host/ProxyMetadataManager.cs [hunk 26]
```diff
@@ -119,7 +119,7 @@ private ProxyMetadataInfo LoadFunctionMetadata()
                     proxyMetadata.Bindings.Add(bindingMetadata);
 
                     proxyMetadata.Name = proxyName;
-                    proxyMetadata.ScriptType = ScriptType.Unknown;
+                    proxyMetadata.Language = "Unknown";
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ProxyMetadataManager.cs]: How is this "Unknown" value used? I wonder if rather than setting it, we just leave it null, which can then signal unknown?

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 27]
```diff
@@ -256,28 +253,27 @@ internal static string DeterminePrimaryScriptFile(JObject functionConfig, string
...
             switch (extension)
             {
                 case "csx":
                 case "cs":
-                    return ScriptType.CSharp;
-                case "js":
-                    return ScriptType.Javascript;
-                case "ts":
-                    return ScriptType.TypeScript;
+                    return "CSharp";
```
Comment (mathewc) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: So you're first parsing the built ins, then falling back to language service?

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 28]
```diff
@@ -256,28 +253,27 @@ internal static string DeterminePrimaryScriptFile(JObject functionConfig, string
...
                 case "fsx":
-                    return ScriptType.FSharp;
+                    return "FSharp";
```
Comment (mathewc) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: Can you use your new enum / constant class?

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs (hunk 27 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Host/FunctionMetadataManager.cs]: Yes!

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs (hunk 28 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Host/FunctionMetadataManager.cs]: done!

File: src/WebJobs.Script/Host/ProxyMetadataManager.cs (hunk 26 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Host/ProxyMetadataManager.cs]: done!

File: src/WebJobs.Script/Description/ScriptTypes.cs (hunk 25 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Description/ScriptTypes.cs]: deleted!

File: src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilationService.cs (hunk 24 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Description/Node/Compilation/TypeScript/TypeScriptCompilationService.cs]: cc: @mhoeger 
NodeJS worker needs to handle this.

File: src/WebJobs.Script/Rpc/LanguageWorkerService.cs (hunk 22 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/LanguageWorkerService.cs]: Added LanguageWorkerConfigOptions to get the list of workerConfigs. This list is also used by FunctionDispatcher to set up invocation subscriptions for on indexed functions

File: src/WebJobs.Script/Rpc/LanguageWorkerService.cs (hunk 21 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/LanguageWorkerService.cs]: Fixed this. Added LanguageWorkerConfigOptions

File: src/WebJobs.Script/Config/ScriptHostOptionsSetup.cs (hunk 20 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Config/ScriptHostOptionsSetup.cs]: Removed _languageWorkerService reference from scriptHostOptions

File: src/WebJobs.Script/Rpc/LanguageWorkerService.cs (hunk 19 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/LanguageWorkerService.cs]: Still figuring this out. Logger is still null.

File: src/WebJobs.Script/Rpc/Configuration/LanguageWorkerConfigOptions.cs [hunk 29]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.WebJobs.Script.Rpc
+{
+    public class LanguageWorkerConfigOptions
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/Configuration/LanguageWorkerConfigOptions.cs]: @brettsam - can you please check this?

File: src/WebJobs.Script/Description/DotNet/Compilation/DotNetCompilationServiceFactory.cs (hunk 23 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Description/DotNet/Compilation/DotNetCompilationServiceFactory.cs]: Perhaps file type?

File: src/WebJobs.Script/Description/DotNet/Compilation/DotNetCompilationServiceFactory.cs [hunk 30]
```diff
@@ -53,19 +53,19 @@ internal static void SetOptimizationLevel(OptimizationLevel? level)
...
         }
 
-        public ICompilationService<IDotNetCompilation> CreateService(ScriptType scriptType, IFunctionMetadataResolver metadata)
+        public ICompilationService<IDotNetCompilation> CreateService(string language, IFunctionMetadataResolver metadata)
         {
-            switch (scriptType)
+            switch (language)
             {
-                case ScriptType.CSharp:
+                case "CSharp":
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/Compilation/DotNetCompilationServiceFactory.cs]: Why not use your new DotNetScriptTypes constants here?

File: src/WebJobs.Script/Rpc/Configuration/LanguageWorkerConfigOptions.cs (hunk 29 above)
Comment (mathewc) [inline src/WebJobs.Script/Rpc/Configuration/LanguageWorkerConfigOptions.cs]: I'd rename this LanguageWorkerOptions to make it more extensible in the future. It's not just about WorkerConfigs, and ConfigOptions is kinda redundant :) Then rename the below property to just WorkerConfigs.

File: src/WebJobs.Script/Rpc/Configuration/LanguageWorkerConfigOptionsSetup.cs [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+{
+    internal class LanguageWorkerConfigOptionsSetup : IConfigureOptions<LanguageWorkerConfigOptions>
+    {
+        private IEnumerable<WorkerConfig> _languageWorkerConfigs;
+        private IConfiguration _configuration;
+
+        public LanguageWorkerConfigOptionsSetup(IConfiguration configuration, ILoggerFactory loggerFactory)
+        {
+            _configuration = configuration;
+            ILogger logger = loggerFactory.CreateLogger("Host.LanguageWorkerConfigOptionsSetup");
```
Comment (mathewc) [inline src/WebJobs.Script/Rpc/Configuration/LanguageWorkerConfigOptionsSetup.cs]: I think all the actual work you're doing here should be moved from the ctor to the Configure method below.

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 32]
```diff
@@ -122,7 +123,9 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
                 // Configuration
                 services.AddSingleton<IOptions<ScriptApplicationHostOptions>>(new OptionsWrapper<ScriptApplicationHostOptions>(applicationHostOptions));
                 services.ConfigureOptions<ScriptHostOptionsSetup>();
-                services.ConfigureOptions<HostHealthMonitorOptionsSetup>();
+
+                // TODO: pgopa only add this to WebHostServiceCollection
```
Comment (mathewc) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: What is this TODO for? Remove this?

File: test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/NodeEndToEndTests.cs [hunk 33]
```diff
@@ -84,7 +84,7 @@ public async Task BlobTriggerToBlobTest()
             Guid.Parse(invocationId);
         }
 
-        [Fact]
+        [Fact (Skip = "Investigate test hang")]
```
Comment (mathewc) [inline test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/NodeEndToEndTests.cs]: You're not going to check it in like this, right?

Comment (pragnagopa) [reply test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/NodeEndToEndTests.cs]: Investigating test failure.  Will undo these changes once I figure it out. Udpated title on the PR to include "Do not merge"
